custom/icds_reports/static/js/directives/download/download.directive.js

```diff
@@ -194,6 +194,31 @@ function DownloadController($rootScope, $location, locationHierarchy, locationsS
         vm.selectedLocations = new Array(maxLevel);
     };
 
+    vm.userHaveAccessToAllLocations = function (locations) {
+        var haveAccessToAllLocationsForLevel = true;
+        window.angular.forEach(locations, function (location) {
+            if (!location.user_have_access) {
+                haveAccessToAllLocationsForLevel = false;
+            }
+        });
+        return haveAccessToAllLocationsForLevel;
+    };
+
+    vm.userLocationIdIsNull = function () {
+        return ["null", "undefined"].indexOf(vm.userLocationId) !== -1;
+    };
+
+    vm.isUserLocationIn = function (locations) {
+        var userLocationInSorted = _.filter(locations, function (location) {
+            return allUserLocationId.indexOf(location.location_id) !== -1;
+        });
+        return userLocationInSorted.length > 0;
+    };
+
+    vm.preventShowingAllOption = function (locations) {
+        return ((!vm.userLocationIdIsNull() && !vm.userHaveAccessToAllLocations(locations)) || vm.isUserLocationIn(locations)) && !haveAccessToAllLocations;
```

Comment:
this is quite hard to reason about since it's negating and then represents the option to NOT show all. 

I think it would be simpler to reason about if it was `vm.showAllOption`

What's the difference between `haveAccessToAllLocations` and `vm.userHaveAccessToAllLocations(locations)`

Also you could out `haveAccessToAllLocations` first in the list to avoid having to evaluate all the other conditions if it's true

Reply:
hey, `haveAccessToAllLocations` is the role permission which tells enough if user is assigned specific location. Can he access all other locations as well or not.
userHaveAccessToAllLocations is checking whether user have access to all Ancestor&descendent locations of the selected location(`locations` argument is the list of those locations sent by backend)

Reply:
Sure Let me simplify this

Reply:
i agree with these points, though i think all of this was just copy/pasted from: https://github.com/dimagi/commcare-hq/blob/master/custom/icds_reports/static/js/directives/location-filter/location-filter.directive.js#L268-L291

Though looking closer at this - why is the logic (I think) unchanged but the name of the function has been reversed? is that intentional?

Reply:
Yeah, function name I changed because it was actually doing opposite of what it is named.

Reply:
classic dashboard, right there